java/ql/src/Security/CWE/CWE-312/SensitiveStorage.qll

```diff
@@ -243,3 +248,72 @@ class Marshallable extends ClassStore {
     )
   }
 }
+
+/* Holds if the method call is a setter method of `SharedPreferences`. */
+private predicate sharedPreferencesInput(DataFlow::Node sharedPrefs, Expr input) {
+  exists(MethodAccess m |
+    m.getMethod() instanceof SharedPreferencesSetMethod and
+    input = m.getArgument(1) and
+    sharedPrefs.asExpr() = m.getQualifier()
+  )
+}
+
+/* Holds if the method call is the store method of `SharedPreferences`. */
+private predicate sharedPreferencesStore(DataFlow::Node sharedPrefs, Expr store) {
+  exists(MethodAccess m |
+    m.getMethod() instanceof SharedPreferencesStoreMethod and
+    store = m and
+    sharedPrefs.asExpr() = m.getQualifier()
+  )
+}
+
+/* Flow from `SharedPreferences` to the method call changing its value. */
+class SharedPreferencesFlowConfig extends TaintTracking::Configuration {
+  SharedPreferencesFlowConfig() { this = "SensitiveStorage::SharedPreferencesFlowConfig" }
+
+  override predicate isSource(DataFlow::Node src) {
+    src.asExpr() instanceof SharedPreferencesEditor
+  }
+
+  override predicate isSink(DataFlow::Node sink) {
+    sharedPreferencesInput(sink, _) or
+    sharedPreferencesStore(sink, _)
+  }
+
+  override predicate isSanitizer(DataFlow::Node n) {
+    exists(MethodAccess ma |
+      ma.getMethod().getName().toLowerCase().matches("%encrypt%") and
+      n.asExpr() = ma.getAnArgument()
+    )
+  }
+}
+
+/** The call to get a `SharedPreferences.Editor` object, which can set shared preferences or be stored to device. */
+class SharedPreferencesEditor extends MethodAccess {
+  SharedPreferencesEditor() {
+    this.getMethod() instanceof SharedPreferencesGetEditorMethod and
+    not exists(
+      MethodAccess cma // not exists `SharedPreferences sharedPreferences = EncryptedSharedPreferences.create(...)`
+    |
+      cma.getQualifier().getType() instanceof TypeEncryptedSharedPreferences and
```

Comment:
Is this quite brittle? Does it depend on the get-editor method acting directly on EncryptedSharedPreferences.create(...), or can it cope with intermediate dataflow, either within a function or across function boundaries?

Reply:
Thanks for the advice. Yes, it's not that robust.

I've revamped the query to use a flow configuration and added two new test cases.